File: node-js/parser.js
import fs from 'fs';
import readline from 'readline';

 async function processBindFile(filePath='zones/example2.com.zone') {
    const fileStream = fs.createReadStream(filePath);
  
    const rl = readline.createInterface({
      input: fileStream,
      crlfDelay: Infinity
    });
  
    let origin;
    let ttl;
    let previousTtl;
    let previousName;

    let records = [];
    let recordsType = ['SOA', 'NS', 'A', 'CNAME', 'MX'];
    recordsType.forEach(element => {
      records[element] = [];
    })

    let soa = {ttl: 0};
    let soaLine = 0;
    let multiLineSoa = false;
    
    let containsTtl = false;
    let type = 'SOA';
    
    for await (const line of rl) {
      
      if (line.length > 1){
        let commentedLine = false;
        let l = line.trim()
                    .replace(/\t/g, ' ')
                    .replace(/\s+/g, ' ');

        let commentIndex = l.indexOf(';');
        if(commentIndex != -1){
            if (commentIndex != 0) {
              let m = l.split(';');
              l = m[0];
            }
            else{
              commentedLine = true;
            }
            
        }
        
        if (!commentedLine) {
          let splittedLine = l.split(' ');
          switch (splittedLine[0]) {
            case '$ORIGIN':
              origin = splittedLine[1];
              break;
            
            case '$TTL':
              ttl = splittedLine[1];
              break;
            
            case '$INCLUDE':
              break;
            
            default:
              if (splittedLine.includes('SOA')){
                previousName = splittedLine[0];
                soa.name = previousName;
                soa.mname = splittedLine[3];
                soa.rname = splittedLine[4];
                
                if (splittedLine.includes(')')){
                  soa.serial = splittedLine[6]; 
                  soa.refresh = splittedLine[7];
                  soa.retry = splittedLine[8];
                  soa.expire = splittedLine[9];
                  soa.minimum = splittedLine[10];
                  
                  records['SOA'].push(soa);
                }
                else{
                  multiLineSoa= true;
                  soaLine++;
                  
                }
                
              }

              recordsType.forEach(element => {
                if (splittedLine.includes(element)){
                  
                  type = element;
                  if (type != 'SOA') {
                  

                    let rr;
                    [rr, previousName, previousTtl] = processRr(splittedLine, containsTtl, previousTtl, previousName, origin, ttl);
                  
                    records[type].push(rr);
                  }
                }
              })
              break;
          }
          
          if(multiLineSoa){
            switch (soaLine) {
              case 2:
                
                soa.serial = splittedLine[0];
                break;
              case 3:
                soa.refresh = splittedLine[0];
                break;
              case 4:
                soa.retry = splittedLine[0];
                break;
              case 5:
                soa.expire = splittedLine[0];
                break;
              case 6:
                soa.minimum = splittedLine[0];
                break;
              default:
                break;
            }
            if(splittedLine.includes(')')){
              records['SOA'].push(soa)
              multiLineSoa = false;
            }
            soaLine++;
          }
        }
      }
    }
    
    return records;
  }

function processRr(splittedLine , containsTtl, previousTtl, previousName, origin, ttl) {
  let rr = {};
  let totalLength = splittedLine.length;
  let isMx = Number(splittedLine[totalLength -2]); 

    switch (totalLength) {
      case 5:
        for (let index = 0; index < totalLength; index++) {
          const element = splittedLine[index];
          if (!element.includes('.')) {
            if(parseInt(element)){
              if(!isMx){
                containsTtl = true;
                previousTtl = element;
                splittedLine.splice(index, 1);
              }
              break;
            }
          } 
          
          
        }
        
        if (!isMx) {
          previousName =splittedLine[0];
          rr.class = splittedLine[1];
          rr.type = splittedLine[2];
          rr.data = splittedLine[3];
        }
        
        break;
      case 4:
        
        for (let index = 0; index < totalLength; index++) {
          const element = splittedLine[index];
          if (!element.includes('.')) {
            if(parseInt(element)){
              if(!isMx){  
                containsTtl = true;
                previousTtl = element;
                splittedLine.splice(index, 1);
              }
              break;
            }
          } 
          
          
        }
        
        if(containsTtl){ //Name is missing
          rr.class = splittedLine[0];
          rr.type = splittedLine[1];
          rr.data = splittedLine[2]; 

        }
        else{
          if(isMx){
            previousName = "@";
            rr.class = splittedLine[0];
            rr.type = splittedLine[1];
            rr.preference = splittedLine[2];
            rr.data = splittedLine[3];
            
          }else{
            previousName = splittedLine[0];
            rr.class = splittedLine[1];
            rr.type = splittedLine[2];
            rr.data = splittedLine[3];
          }
          
        }
        
        break;
      case 3:
        rr.class = splittedLine[0];
        rr.type = splittedLine[1];
        rr.data = splittedLine[2];
        
        break;
      case 2:
        break; 
      default:
        break;
    }
    rr.name = previousName || origin;
    rr.ttl =  previousTtl || ttl;

    return [rr, previousName, previousTtl];
}

export {processBindFile};

File: ruby/parser.rb
class Parser
    def self.process_rr(splitted_line, contains_ttl, previous_ttl, previous_name, origin, ttl)

        rr = {}
        total_length = splitted_line.length
        isMx = Integer(splitted_line[total_length - 2], exception: false)

        case total_length
        when 5
            
        when 4
            for index in 0..total_length do
                element = splitted_line[index]
                if !element.nil?
                    if !element.include?('.')
                        if Integer(element, exception: false)
                            if !isMx
                                contains_ttl = true
                                previous_ttl = element
                                splitted_line.delete_at(index)
                            end
                        end
                    end
                end
            end
            if contains_ttl
                rr[:class] = splitted_line[0]
                rr[:type] = splitted_line[1]
                rr[:data] = splitted_line[2]
            else
                if isMx
                    previous_name = '@'
                    rr[:class] = splitted_line[0]
                    rr[:type] = splitted_line[1]
                    rr[:preference] = splitted_line[2]
                    rr[:data] = splitted_line[3]
                else
                    previous_name = splitted_line[0]
                    rr[:class] = splitted_line[1]
                    rr[:type] = splitted_line[2]
                    rr[:data] = splitted_line[3]
                end
            end
        when 3
            rr[:class] = splitted_line[0]
            rr[:type] = splitted_line[1]
            rr[:data] = splitted_line[2] 
        else
            
        end
        rr[:name] = previous_name || origin
        rr[:ttl] = previous_ttl || ttl

        return rr, previous_name, previous_ttl
        
    end
    def self.compute(file_path)
        records = {}
        supported_record_types = ['SOA', 'NS', 'A', 'CNAME', 'MX']
        supported_record_types.each  {|record_type| records[record_type] = []}

        soa = {
            ttl: 0,
            mname: '',
            rname: '',
            serial: 0,
            refresh: 0,
            retry: 0,
            expire: 0,
            minimum: 0
        }
        soa_line_count = 0
        multi_line_soa = false

        contains_ttl = false
        type = 'SOA'

        origin = ''
        ttl = 0

        previous_name = ''
        previous_ttl = 0

        File.open(file_path).each do |line|
            commented_line = false
            l = line.chomp
                    .gsub(/\t/, ' ')
                    .gsub(/\s+/, ' ')
            if l.length > 1
                

                comment_index = l.index(';')

                if !comment_index.nil?
                    if comment_index != 0
                        m = l.split(';')
                        l = m[0]
                    else
                        commented_line = true
                    end
                
                end

                if !commented_line
                    
                    splitted_line = l.split(' ')
                    
                    case splitted_line[0]
                    when '$ORIGIN'
                        origin = splitted_line[1]
                        
                    when '$TTL' 
                        ttl = splitted_line[1]
                    when '$INCLUDE'
                        # TODO
                    else
                        if splitted_line.include?('SOA')
                            previous_name = splitted_line[0]
                            soa[:name] = previous_name
                            soa[:mname] = splitted_line[3]
                            soa[:rname] = splitted_line[4]
                            if splitted_line.include?(')')
                                
                                soa[:serial] = splitted_line[6]
                                soa[:refresh] = splitted_line[7]
                                soa[:retry] = splitted_line[8]
                                soa[:expire] = splitted_line[9]
                                soa[:minimum] = splitted_line[10]
                                
                                records['SOA'].push(soa)
                            else
                                multi_line_soa = true
                                soa_line_count = soa_line_count + 1
                                
                            end
                        
                        end

                        supported_record_types.each do |element|
                            if splitted_line.include?(element)
                                type = element
                                if type != 'SOA'
                                    rr, previous_name, previous_ttl = process_rr(splitted_line, contains_ttl, previous_ttl, previous_name,
                                    origin, ttl)
                                    records[type].push(rr)
                                end
                            end
                        end
                    end

                    if multi_line_soa
                        case soa_line_count
                        when 2
                            soa[:serial] = splitted_line[0]
                            
                        when 3
                            soa[:refresh] = splitted_line[0] 
                        when 4
                            soa[:retry] = splitted_line[0] 

                        when 5
                            soa[:expire] = splitted_line[0] 
                        when 6
                            soa[:minimum] = splitted_line[0]     
                        else
                            
                        end
                        if splitted_line.include?(')')
                            multi_line_soa = false
                            records['SOA'].push(soa)
                        end
                    end
                end
            end
            
        end
        return records
    end    # return records

end